llvm/test/CodeGen/SPIRV/optimizations/recognize-popcnt-loop.ll

```diff
@@ -0,0 +1,114 @@
+; RUN: opt -O3 -mtriple=spirv32-- %s -o - | llc -O3 -mtriple=spirv32-- -o - | FileCheck %s
+; RUN: %if spirv-tools %{ opt -O3 -mtriple=spirv32-- %s -o - | llc -O3 -mtriple=spirv32-- -o - -filetype=obj | spirv-val %}
```

Comment:
As a nit, it'd be useful to run this as `llc -verify-machineinstrs ...`.

Reply:
This is actually a great idea, because it tickles something [that looks like a notable bug](https://github.com/llvm/llvm-project/blob/f7dfaf35065fc858056a206b10c3dfbf8029f801/llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp#L355) - that insertion is incorrect, we should skip over PHIs if the successor is a PHI. However, doing that doesn't quite solve it, as vreg lifetimes get messed up and verification still fails. I'll need to spend a bit more time to understand the transform. This can be verified, independently from, this patch, by using  `-verify-machineinstrs` on e.g. `/llvm/test/CodeGen/SPIRV/phi-spvintrinsic-dominate.ll`. So this doesn't make things worse - we could merge it and deal with the issue separately, I reckon.

Reply:
> that insertion is incorrect

yes, this has already been addressed in https://github.com/llvm/llvm-project/pull/109660 (https://github.com/llvm/llvm-project/blob/01c777628e17fce039f1b50297e6c2125445a255/llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp#L344-L366)

> vreg lifetimes get messed up and verification still fails

I didn't see the case to be sure, but I'd guess that it's the same problem as we're discussing currently within https://github.com/llvm/llvm-project/pull/110019